CMakeLists.txt

```diff
@@ -34,23 +35,80 @@ if(FINUFFT_BUILD_MATLAB)
 
 # Find required packages
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
-find_package(FFTW REQUIRED)
-
-
-set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
 
+if (NOT FINUFFT_FFTW_LIBRARIES STREQUAL DEFAULT)
+    find_package(FFTW)
+    if (NOT FFTW_FOUND)
+        # USING CPM TO HANDLE DEPENDENCIES
+        set(CPM_DOWNLOAD_VERSION 0.34.0)
+        if (CPM_SOURCE_CACHE)
+            set(CPM_DOWNLOAD_LOCATION "${CPM_SOURCE_CACHE}/cpm/CPM_${CPM_DOWNLOAD_VERSION}.cmake")
+        elseif (DEFINED ENV{CPM_SOURCE_CACHE})
+            set(CPM_DOWNLOAD_LOCATION "$ENV{CPM_SOURCE_CACHE}/cpm/CPM_${CPM_DOWNLOAD_VERSION}.cmake")
+        else ()
+            set(CPM_DOWNLOAD_LOCATION "${CMAKE_BINARY_DIR}/cmake/CPM_${CPM_DOWNLOAD_VERSION}.cmake")
+        endif ()
+
+        if (NOT (EXISTS ${CPM_DOWNLOAD_LOCATION}))
+            message(STATUS "Downloading CPM.cmake to ${CPM_DOWNLOAD_LOCATION}")
+            file(DOWNLOAD
+                    https://github.com/TheLartians/CPM.cmake/releases/download/v${CPM_DOWNLOAD_VERSION}/CPM.cmake
+                    ${CPM_DOWNLOAD_LOCATION}
+                    )
+        endif ()
+
+        include(${CPM_DOWNLOAD_LOCATION})
+
+        set(FFTW_VERSION 3.3.10)
+        if (FINUFFT_FFTW_SUFFIX STREQUAL THREADS)
+            set(FINUFFT_USE_THREADS ON)
+        else ()
+            set(FINUFFT_USE_THREADS OFF)
+        endif ()
+        CPMAddPackage(
+                NAME fftw3
+                OPTIONS "ENABLE_AVX2 ON" "BUILD_TESTS OFF" "BUILD_SHARED_LIBS OFF" "ENABLE_THREADS ${FINUFFT_USE_THREADS}" "ENABLE_OPENMP ${FINUFFT_USE_OPENMP}"
+                URL "http://www.fftw.org/fftw-${FFTW_VERSION}.tar.gz"
+                URL_HASH "MD5=8ccbf6a5ea78a16dbc3e1306e234cc5c"
+                EXCLUDE_FROM_ALL YES
+                GIT_SHALLOW YES
+        )
+
+        CPMAddPackage(
+                NAME fftw3f
+                OPTIONS "ENABLE_AVX2 ON" "BUILD_TESTS OFF" "BUILD_SHARED_LIBS OFF" "ENABLE_FLOAT ON" "ENABLE_THREADS ${FINUFFT_USE_THREADS}" "ENABLE_OPENMP ${FINUFFT_USE_OPENMP}"
+                URL "http://www.fftw.org/fftw-${FFTW_VERSION}.tar.gz"
+                URL_HASH "MD5=8ccbf6a5ea78a16dbc3e1306e234cc5c"
+                EXCLUDE_FROM_ALL YES
+                GIT_SHALLOW YES
+        )
+        set(FINUFFT_FFTW_LIBRARIES fftw3 fftw3f)
+        if (FINUFFT_USE_THREADS)
+            list(APPEND FINUFFT_FFTW_LIBRARIES fftw3_threads fftw3f_threads)
+        elseif (FINUFFT_USE_OPENMP)
+            list(APPEND FINUFFT_FFTW_LIBRARIES fftw3_omp fftw3f_omp)
+        endif ()
+
+        foreach (element IN LISTS FINUFFT_FFTW_LIBRARIES)
+            set_property(TARGET ${element} PROPERTY POSITION_INDEPENDENT_CODE ON)
+        endforeach ()
+
+    else ()
+        set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+    endif ()
+endif ()
```

Comment:
Thanks for the refinement @DiamonDinoia. One minor suggestion, this fetching fftw part may be factored to cmake directory and using include(cmake/FFTW.cmake)? BTW I mostly use native CMake FetchContent, is CPM.cmake more commonly used?

Reply:
We can move this in another file if that's preferable.

Not sure what is more commonly used but CPM is a wrapper around FetchContent that makes things much easier. It is also offer the possibility of caching dependencies between multiple projects. In general, I find it more convenient. Especially, using FetchContent to specify options and exclude binaries from install is cumbersome

Reply:
Actually, I think we should download the cmake module using cpm too instead of copy pasting and external file that is available on git. I will make the changes once I have time.